Initial commit for Help command (doesn't compile)

artem-kim opened the pull request:
Hi team, I am pulling my hair out a bit with trying to make the code run. I added a "HelpCommand.java" to the commands folder (with some dummy method there). Could somebody have a look at the errors and help me troubleshoot it? 

[ERROR] /Users/wwhd/Documents/SE2/newbank3/newbank/src/main/java/newbank/server/NewBankClientHandler.java:[52,26] incompatible types: invalid constructor reference
[ERROR]     constructor HelpCommand in class newbank.server.commands.HelpCommand cannot be applied to given types
[ERROR]       required: no arguments
[ERROR]       found:    newbank.server.NewBank,java.lang.String[],newbank.server.CustomerID
[ERROR]       reason: actual and formal argument lists differ in length

artem-kim committed:
42 - Add HELP command implementation

michowski commented:
Hi Artem, the IDE/editor should show you:

```
The type HelpCommand does not define HelpCommand(NewBank, String[], CustomerID) that is applicable hereJava(603979903)
```

in `src/main/java/newbank/server/NewBankClientHandler.java:52`

michowski commented:
@artem-kim I've added and pushed the missing constructor. Could you please do `git pull origin 42-ui-update-v2`?

artem-kim commented:
@michowski @juliano60 @ZacTredger @claudist good morning team, I finished the HelpCommand, added welcome instructions and updated the protocol with a new help command. Please review.

juliano60 commented on src/main/java/newbank/server/NewBankClientHandler.java:
```diff
@@ -78,6 +80,7 @@ public void run() {
         out.println(String.format("Received request [%s]", request));
 
         hasMore = processRequest(request);
+        out.println("Enter 'HELP' to see all commands");
```
This needs to move to the client side.

Regards,
Jules

juliano60 replied:
Fixed.

Regards,
Jules

juliano60 commented on src/main/java/newbank/client/ExampleClient.java:
```diff
@@ -111,6 +112,11 @@ private void displayWelcomingBanner() {
   }
 
+  private void welcomeMenu() {
+    String menu = "Please login to your account ('LOGIN username password') or register ('REGISTER username password')";
```
Please catch up with main as @claudist has already checked that in.

Regards,
Jules

juliano60 replied:
Resolved.

juliano60 commented on src/main/java/newbank/server/commands/HelpCommand.java:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  @Override
+  public String execute() {
```
Rather than displaying the syntax name, which they can get by issuing "COMMAND HELP", we could display,
"COMMAND  a brief explanation of the command".
Try "git help" on the command line and use the output for inspiration.

Regards,
Jules

juliano60 commented on docs/protocol.md:
```diff
@@ -68,6 +68,33 @@
+  </tr>
+  <tr>
+    <td><strong>Example</strong></td>
+    <td colspan="2"><code>$ HELP</code><br /><code>You have the following commands available:</code><br /><code>...................GENERAL...................</code><br /><code>LOGIN username password</code><br /><code>QUIT</code><br /><code>............ACCOUNT ADMINISTRATION............</code><br /><code>DEFAULT account_name</code><br /><code>NEWACCOUNT name [DEFAULT]</code><br /><code>REGISTER u ...
```
The example does not make sense for this command. The output is too long and not an example as it doesn't change.
Note that it says that you need to be logged in in the output which is incorrect.
The command also should return "SUCCESS or FAIL" probably.

Kindest Regards,
Jules

juliano60 reviewed (changes requested):
Good attempt @artem-kim , a few comments to address.

Kindest Regards,
Jules

juliano60 commented:
@artem-kim I resolved some of the issues (caught your branch up with "main" and moved some stuffs to the client side).
Still some outstanding issues for you to look at. Just catch up with origin/42-ui-update-v2 to download the changes.

Kindest Regards,
Jules

juliano60 commented:
@michowski  @ZacTredger Could you please review?
= tidied up the HELP command
= added fixes for SHOWCCOUNTS

Actually forgot something, hold off the review for now...

Regards,
Jules

juliano60 commented:
![image](https://user-images.githubusercontent.com/8995677/114750170-9605c280-9d4b-11eb-99ad-d709fc777975.png)

ZacTredger commented on docs/protocol.md:
```diff
@@ -68,6 +68,33 @@
+  </tr>
+  <tr>
+    <td><strong>Example</strong></td>
+    <td colspan="2"><code>$ HELP</code> SUCCESS ...</code></td>
```
You close the `code` tag twice. Better to make sure though, eh?

ZacTredger commented on docs/protocol.md:
```diff
@@ -68,6 +68,33 @@
+  </tr>
+  <tr>
+    <td><strong>Returns</strong></td>
+    <td><em>info</em></td>
```
Should this be `SUCCESS` instead of `info`?

ZacTredger commented on src/main/java/newbank/utils/Config.java:
```diff
@@ -0,0 +1,5 @@
+package newbank.utils;
+
+public class Config {
+	public static final String MULTILINE_INFO_SEPARATOR = "~~";
```
This was a great idea :+1:

ZacTredger commented on src/test/java/newbank/TestApp.java:
```diff
@@ -126,11 +130,13 @@ public void canMoveMoney() throws IOException {
-    accountSummary = display.getLine();
+    accountSummary = info[2];
     assertThat(accountSummary, matchesPattern("Savings:\\s+201.19\\s+GBP"));
-    assertThat(display.getLine(), equalTo(""));
```
I feel like this pattern is repeatedly crying out "extract method!" No?

ZacTredger reviewed (dismissed):
Added some teeny tiny suggestions. Only the `info` vs `SUCCESS` thing in `protocol.md` is actually observable to users.
I'm approving, but if you want, I can just add another commit to fix these little things? Let me know.
Great work guys
Z
